### PR TITLE
Optimize activity registry descriptor lookup

### PR DESCRIPTION
**Pull request by sfmskywalker**
## Summary

Implements the highest-impact, lowest-effort opportunity from the Elsa Core performance scan: `ActivityRegistry.Find(string type)` now uses a per-registry latest descriptor index instead of scanning `ActivityDescriptors.Values.Where(...).MaxBy(...)` for every lookup.

The index preserves tenant-specific precedence over agnostic descriptors and is kept consistent across add, remove, clear, provider clear, tenant clear, and provider refresh paths. Provider refresh now also removes descriptors from registries that a provider no longer contributes to, so stale latest-by-type entries cannot survive when a provider drops a tenant group.

## Performance opportunities from the scan

1. `ActivityRegistry.Find(string type)` latest-version descriptor lookup scanned all descriptors on hot paths. This PR implements O(1) latest-by-type lookup.
2. Descriptor removal needed targeted latest-version recomputation instead of repeated full scans during lookup.
3. Provider refresh paths could leave stale registry/provider entries when a provider stopped returning descriptors for a tenant group.
4. Tenant-specific descriptor precedence should avoid comparing tenant descriptors against agnostic descriptors when a tenant match exists.
5. Repeated LINQ allocation in registry lookup paths should be avoided for exact type lookups.
6. Registry mutation paths should update indexes incrementally and only recompute for affected type names.
7. Clear-provider flows should invalidate lookup indexes alongside primary descriptor dictionaries.
8. Registry clear/tenant clear operations should drop indexed state with the owning registry data.
9. Remaining predicate-based descriptor searches still scan by design and are candidates for future workload-specific indexes.
10. Broader workflow compilation/execution lookup paths should be measured next after removing this descriptor lookup bottleneck.

## Tests

Not run locally: `dotnet` is not available on PATH in this environment (`dotnet: command not found`).

### CODE REVIEW

File: src/modules/Elsa.Workflows.Core/Services/ActivityRegistry.cs [hunk 1]
```diff
@@ -197,34 +193,47 @@ public async Task RefreshDescriptorsAsync(IActivityProvider activityProvider, Ca
...
+        foreach (var registry in GetRegistriesWithProvider(providerType).Where(x => !refreshedRegistries.Contains(x)))
+        {
+            if (!registry.ProvidedActivityDescriptors.TryRemove(providerType, out var oldDescriptors))
+                continue;
+
+            foreach (var oldDescriptor in oldDescriptors.ToList())
+            {
+                RemoveDescriptor(registry, oldDescriptor);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/modules/Elsa.Workflows.Core/Services/ActivityRegistry.cs]: Leaving this guard in place. GetRegistriesWithProvider already filters registries that have this provider, but TryRemove can still fail if the concurrent dictionary changes and the removed descriptor list is needed for the cleanup loop. Moving this into a Where filter would either duplicate the lookup or obscure the out value.

File: src/modules/Elsa.Workflows.Core/Services/ActivityRegistry.cs [hunk 2]
```diff
@@ -305,6 +319,47 @@ private ICollection<ActivityDescriptor> GetOrCreateProviderDescriptors(TenantReg
...
+    private IEnumerable<TenantRegistryData> GetRegistriesWithProvider(Type providerType)
+    {
+        if (_agnosticRegistry.ProvidedActivityDescriptors.ContainsKey(providerType))
+            yield return _agnosticRegistry;
+
+        foreach (var registry in _tenantRegistries.Values)
+        {
+            if (registry.ProvidedActivityDescriptors.ContainsKey(providerType))
+                yield return registry;
+        }
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/modules/Elsa.Workflows.Core/Services/ActivityRegistry.cs:334]: Addressed in 6d1b7e321 by filtering _tenantRegistries.Values directly with Where in GetRegistriesWithProvider. Validation: dotnet test test/unit/Elsa.Workflows.Core.UnitTests/Elsa.Workflows.Core.UnitTests.csproj --filter ActivityRegistryTests -p:CollectCoverage=false (20 passed).

File: src/modules/Elsa.Workflows.Core/Services/ActivityRegistry.cs [hunk 3]
```diff
@@ -305,6 +319,47 @@ private ICollection<ActivityDescriptor> GetOrCreateProviderDescriptors(TenantReg
...
+    {
+        var latestDescriptor = registry.ActivityDescriptors.Values
+            .Where(x => x.TypeName == typeName)
+            .MaxBy(x => x.Version);
+
+        if (latestDescriptor == null)
+            registry.LatestActivityDescriptors.TryRemove(typeName, out _);
+        else
+            registry.LatestActivityDescriptors[typeName] = latestDescriptor;
+    }
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/modules/Elsa.Workflows.Core/Services/ActivityRegistry.cs:370]: Addressed in 6d1b7e321 by replacing the LINQ Where/MaxBy recomputation with a simple loop over the registry descriptors. Validation: dotnet test test/unit/Elsa.Workflows.Core.UnitTests/Elsa.Workflows.Core.UnitTests.csproj --filter ActivityRegistryTests -p:CollectCoverage=false (20 passed).

File: src/modules/Elsa.Workflows.Core/Services/ActivityRegistry.cs [hunk 4]
```diff
@@ -254,6 +267,7 @@ public void Clear()
     {
         _tenantRegistries.Clear();
         _agnosticRegistry.ActivityDescriptors.Clear();
+        _agnosticRegistry.LatestActivityDescriptors.Clear();
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/modules/Elsa.Workflows.Core/Services/ActivityRegistry.cs:273]: Addressed in da4498398 by clearing _manualActivityDescriptors in Clear() and adding a regression test that verifies GetDescriptorsAsync(), ListAll(), and Find() are empty after clearing a registry populated through RegisterAsync. Validation: dotnet test test/unit/Elsa.Workflows.Core.UnitTests/Elsa.Workflows.Core.UnitTests.csproj --filter ActivityRegistryTests -p:CollectCoverage=false (21 passed).

File: src/modules/Elsa.Workflows.Core/Services/ActivityRegistry.cs [hunk 5]
```diff
@@ -305,6 +319,52 @@ private ICollection<ActivityDescriptor> GetOrCreateProviderDescriptors(TenantReg
...
+    {
+        ActivityDescriptor? latestDescriptor = null;
+        foreach (var descriptor in registry.ActivityDescriptors.Values)
+        {
+            if (descriptor.TypeName != typeName)
+                continue;
+
+            if (latestDescriptor == null || descriptor.Version > latestDescriptor.Version)
+                latestDescriptor = descriptor;
+        }
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/modules/Elsa.Workflows.Core/Services/ActivityRegistry.cs:364]: Keeping the explicit loop here. This path was changed specifically to avoid the LINQ Where/MaxBy allocation noted by Copilot on the previous review; moving the TypeName filter back into Where would reintroduce part of that overhead. The loop is intentional and covered by ActivityRegistryTests.

File: src/modules/Elsa.Workflows.Core/Services/ActivityRegistry.cs [hunk 6]
```diff
@@ -305,6 +320,52 @@ private ICollection<ActivityDescriptor> GetOrCreateProviderDescriptors(TenantReg
...
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/modules/Elsa.Workflows.Core/Services/ActivityRegistry.cs:364]: Declining this suggestion for the same reason as the previous thread: RecomputeLatestDescriptor intentionally uses a simple loop to avoid LINQ iterator allocation in this index-maintenance path. Moving the filter into Where would reintroduce that overhead.

File: src/modules/Elsa.Workflows.Core/Services/ActivityRegistry.cs [hunk 7]
```diff
@@ -305,6 +320,52 @@ private ICollection<ActivityDescriptor> GetOrCreateProviderDescriptors(TenantReg
...
+    private IEnumerable<TenantRegistryData> GetRegistriesWithProvider(Type providerType)
+    {
+        if (_agnosticRegistry.ProvidedActivityDescriptors.ContainsKey(providerType))
+            yield return _agnosticRegistry;
+
+        foreach (var registry in _tenantRegistries.Values.Where(x => x.ProvidedActivityDescriptors.ContainsKey(providerType)))
+        {
+            yield return registry;
+        }
+    }
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/modules/Elsa.Workflows.Core/Services/ActivityRegistry.cs:335]: Addressed in dbfe247b6 by changing GetRegistriesWithProvider back to a simple foreach with an in-loop provider check, avoiding the Values.Where(...) iterator allocation. Validation: dotnet test test/unit/Elsa.Workflows.Core.UnitTests/Elsa.Workflows.Core.UnitTests.csproj --filter ActivityRegistryTests -p:CollectCoverage=false (21 passed).

File: src/modules/Elsa.Workflows.Core/Services/ActivityRegistry.cs [hunk 8]
```diff
@@ -166,7 +162,7 @@ public async Task RegisterAsync([DynamicallyAccessedMembers(DynamicallyAccessedM
         var activityDescriptor = await activityDescriber.DescribeActivityAsync(activityType, cancellationToken);
 
         var registry = GetOrCreateRegistry(activityDescriptor.TenantId);
-        Add(activityDescriptor, registry.ActivityDescriptors, _manualActivityDescriptors);
+        Add(activityDescriptor, registry, _manualActivityDescriptors);
         _manualActivityDescriptors.Add(activityDescriptor);
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/modules/Elsa.Workflows.Core/Services/ActivityRegistry.cs]: Addressed in 26017e3b7 by removing the redundant _manualActivityDescriptors.Add call; Add(...) already appends to the passed manual descriptor collection. Validation: dotnet test test/unit/Elsa.Workflows.Core.UnitTests/Elsa.Workflows.Core.UnitTests.csproj --filter ActivityRegistryTests -p:CollectCoverage=false (21 passed).

File: src/modules/Elsa.Workflows.Core/Services/ActivityRegistry.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/modules/Elsa.Workflows.Core/Services/ActivityRegistry.cs]: Addressed in 26017e3b7 by replacing GetRegistriesWithProvider(providerType).Where(...) with a simple foreach plus Contains check inside RefreshDescriptorsAsync. Validation: dotnet test test/unit/Elsa.Workflows.Core.UnitTests/Elsa.Workflows.Core.UnitTests.csproj --filter ActivityRegistryTests -p:CollectCoverage=false (21 passed).

File: src/modules/Elsa.Workflows.Core/Services/ActivityRegistry.cs [hunk 9]
```diff
@@ -305,6 +320,53 @@ private ICollection<ActivityDescriptor> GetOrCreateProviderDescriptors(TenantReg
...
+    private IEnumerable<TenantRegistryData> GetRegistriesWithProvider(Type providerType)
+    {
+        if (_agnosticRegistry.ProvidedActivityDescriptors.ContainsKey(providerType))
+            yield return _agnosticRegistry;
+
+        foreach (var registry in _tenantRegistries.Values)
+        {
+            if (registry.ProvidedActivityDescriptors.ContainsKey(providerType))
+                yield return registry;
+        }
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/modules/Elsa.Workflows.Core/Services/ActivityRegistry.cs:334]: Declining this suggestion because this method is in the provider-refresh/index-maintenance path and Copilot specifically requested avoiding the Values.Where(...) iterator allocation here. The explicit loop is intentional.

File: src/modules/Elsa.Workflows.Core/Services/ActivityRegistry.cs [hunk 10]
```diff
@@ -305,6 +320,53 @@ private ICollection<ActivityDescriptor> GetOrCreateProviderDescriptors(TenantReg
...
+    {
+        ActivityDescriptor? latestDescriptor = null;
+        foreach (var descriptor in registry.ActivityDescriptors.Values)
+        {
+            if (descriptor.TypeName != typeName)
+                continue;
+
+            if (latestDescriptor == null || descriptor.Version > latestDescriptor.Version)
+                latestDescriptor = descriptor;
+        }
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/modules/Elsa.Workflows.Core/Services/ActivityRegistry.cs:364]: Declining this suggestion for the same reason as the related RecomputeLatestDescriptor thread: the explicit loop avoids LINQ iterator allocation in this index-maintenance path, which is the performance goal of this PR.
